src/cargo/ops/registry/publish.rs

```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
-            .finish_hex();
-        let operation = Operation::Publish {
-            name: pkg.name().as_str(),
-            vers: &ver,
-            cksum: &hash,
-        };
-        registry.set_token(Some(auth::auth_token(
-            &opts.gctx,
-            &source_ids.original,
-            None,
-            operation,
-            vec![],
-            false,
-        )?));
-    }
+    let (mut order, mut ready) = PublishOrder::new(&pkg_dep_graph.graph);
+    let mut to_confirm: HashSet<_> = ready.iter().copied().collect();
+    let mut remaining: HashSet<_> = just_pkgs.iter().map(|pkg| pkg.package_id()).collect();
 
-    opts.gctx
-        .shell()
-        .status("Uploading", pkg.package_id().to_string())?;
-    transmit(
-        opts.gctx,
-        ws,
-        pkg,
-        tarball.file(),
-        &mut registry,
-        source_ids.original,
-        opts.dry_run,
-    )?;
-    if !opts.dry_run {
-        const DEFAULT_TIMEOUT: u64 = 60;
-        let timeout = if opts.gctx.cli_unstable().publish_timeout {
-            let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
-            timeout.unwrap_or(DEFAULT_TIMEOUT)
+    while !to_confirm.is_empty() {
+        if !ready.is_empty() {
+            opts.gctx.shell().status(
+                "Uploading",
+                ready.iter().map(PackageId::to_string).sorted().join(", "),
+            )?;
+        }
+
+        for pkg_id in &ready {
+            let (pkg, (_features, tarball)) = &pkg_dep_graph.packages[pkg_id];
+
+            if !opts.dry_run {
+                let ver = pkg.version().to_string();
+
+                tarball.file().seek(SeekFrom::Start(0))?;
+                let hash = cargo_util::Sha256::new()
+                    .update_file(tarball.file())?
+                    .finish_hex();
+                let operation = Operation::Publish {
+                    name: pkg.name().as_str(),
+                    vers: &ver,
+                    cksum: &hash,
+                };
+                registry.set_token(Some(auth::auth_token(
+                    &opts.gctx,
+                    &source_ids.original,
+                    None,
+                    operation,
+                    vec![],
+                    false,
+                )?));
+            }
+
+            transmit(
+                opts.gctx,
+                ws,
+                pkg,
+                tarball.file(),
+                &mut registry,
+                source_ids.original,
+                opts.dry_run,
+            )?;
+
+            if !opts.dry_run {
+                // Short does not include the registry name.
+                let short_pkg_description = format!("{} v{}", pkg.name(), pkg.version());
+                let source_description = source_ids.original.to_string();
+                ws.gctx().shell().status(
+                    "Uploaded",
+                    format!("{short_pkg_description} to {source_description}"),
+                )?;
```

Comment:
At the time, the uploaded message was important context for what phase we are at with waiting.  I wonder if its just extra noise in this scenario.

Reply:
Looking over how this is done, maybe it does make sense.  Unsure